File: apps/web/src/shared/api/problem-details.ts
/**
 * Erro da API no formato Problem Details.
 *
 * Todo erro vindo do backend chega aqui tipado, com `correlation_id` — que é o
 * que o suporte pede quando o usuário relata um problema.
 */
export interface FieldError {
  field: string;
  message: string;
}

export interface ProblemDetails {
  type: string;
  title: string;
  status: number;
  detail: string;
  instance: string;
  correlation_id: string;
  errors: FieldError[];
}

export class ApiError extends Error {
  readonly problem: ProblemDetails;

  constructor(problem: ProblemDetails) {
    super(problem.detail || problem.title);
    this.name = 'ApiError';
    this.problem = problem;
  }

  get status(): number {
    return this.problem.status;
  }

  /** Último segmento do `type`, ex.: "period-closed". */
  get codigo(): string {
    return this.problem.type.split('/').pop() ?? 'unknown';
  }

  get correlationId(): string {
    return this.problem.correlation_id;
  }

  get erroDeCampo(): Record<string, string> {
    return Object.fromEntries(this.problem.errors.map((e) => [e.field, e.message]));
  }
}

export function problemaGenerico(status: number, detail: string): ProblemDetails {
  return {
    type: 'https://rfbalance/errors/network-error',
    title: 'Falha de comunicação',
    status,
    detail,
    instance: '',
    correlation_id: '',
    errors: [],
  };
}

File: apps/api/app/platform/errors/problem_details.py
"""Problem Details (RFC 7807) — formato único de erro da API."""

from __future__ import annotations

from dataclasses import dataclass, field

BASE_URI = "https://rfbalance/errors"


@dataclass(frozen=True, slots=True)
class FieldError:
    field: str
    message: str


@dataclass(frozen=True, slots=True)
class ProblemDetails:
    type: str
    title: str
    status: int
    detail: str
    instance: str
    correlation_id: str
    errors: list[FieldError] = field(default_factory=list)

    @classmethod
    def de_codigo(
        cls,
        *,
        code: str,
        title: str,
        status: int,
        detail: str,
        instance: str,
        correlation_id: str,
        errors: list[FieldError] | None = None,
    ) -> ProblemDetails:
        return cls(
            type=f"{BASE_URI}/{code}",
            title=title,
            status=status,
            detail=detail,
            instance=instance,
            correlation_id=correlation_id,
            errors=errors or [],
        )

    def para_dicionario(self) -> dict[str, object]:
        return {
            "type": self.type,
            "title": self.title,
            "status": self.status,
            "detail": self.detail,
            "instance": self.instance,
            "correlation_id": self.correlation_id,
            "errors": [{"field": e.field, "message": e.message} for e in self.errors],
        }
